File: src/gestorAplicacion/gestorMusica/Lista.java
package gestorAplicacion.gestorMusica;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Set;

import gestorAplicacion.gestorPersonas.*;

public class Lista extends Musica implements Serializable {
	
	private static final long serialVersionUID = 1L;
	private static ArrayList<Lista> listasExistentes;
	private ArrayList<Cancion> lista;
	private Set<Cancion> listaColaborativa;
	protected ArrayList<Usuario> usuarios;
	protected Usuario usuario;
	private String descripcion;
	
	static {
		setListasExistentes(new ArrayList<Lista>());
	}
	
	public Lista(String nombre, Usuario usuario, String descripcion) {
		super(nombre);
		lista = new ArrayList<>();
		this.usuario = usuario;
		this.descripcion = descripcion;
		this.usuario.getColeccion().agregarLista(this);
		listasExistentes.add(this);
	}
	
	public Lista(String nombre, ArrayList<Cancion> lista, Usuario usuario, String descripcion) {
		this(nombre, usuario, descripcion);
		this.lista = lista;
	}
	
	public Lista(String nombre, ArrayList<Usuario> usuarios, Set<Cancion> canciones) {
		super(nombre);
		this.usuarios = usuarios;
		this.listaColaborativa = canciones;
		//ver como afecta que en el constructor solo reciba usuarios invidivuales en el caso de tener arrays 
		this.colaborativa(usuarios);
	}
	
	
	public ArrayList<Cancion> getLista() {
		return lista;
	}

	public void setLista(ArrayList<Cancion> lista) {
		this.lista = lista;
	}

	public Set<Cancion> getListaColaborativa() {
		return listaColaborativa;
	}

	public void setListaColaborativa(Set<Cancion> listaColaborativa) {
		this.listaColaborativa = listaColaborativa;
	}

	public Usuario getUsuario() {
		return usuario;
	}

	public void setUsuario(Usuario usuario) {
		this.usuario = usuario;
	}
	
	public static ArrayList<Lista> getListasExistentes() {
		return listasExistentes;
	}

	public static void setListasExistentes(ArrayList<Lista> listasExistentes) {
		Lista.listasExistentes = listasExistentes;
	}

	public ArrayList<Usuario> getUsuarios() {
		return usuarios;
	}

	public void setUsuarios(ArrayList<Usuario> usuarios) {
		this.usuarios = usuarios;
	}

	public String agregarCancion(Cancion cancion) {
		lista.add(cancion);
		return "Se ha agregado la canción "+cancion.getNombre()+" a la lista "+nombre+" con exito";
	}
	
	public String eliminarCancion(Cancion cancion) {
		lista.remove(cancion);
		return "Se ha eliminado la canción "+cancion.getNombre()+" de la lista "+nombre+"con exito";
	}
	
	@Override
	public String toString() {
		return "Se está reproduciendo la lista " + nombre;
	}
	
	public String infoLista() {
		String des = "";
		if (lista.size() > 0) {
			for (Cancion cancion : lista) {
				des += cancion.descripcion() + "\n";
			}
			return "Lista: " + nombre + "\n" + "\n" + "Canciones: " + "\n" + des;
		} else {
			return "La lista: " + nombre + " está vacía\n";
		}
	}
	
	public String infoColaborativa() {
		String des = "";
		if (listaColaborativa.size() > 0) {
			for (Cancion cancion : listaColaborativa) {
				des += cancion.descripcion() + "\n";
			}
			return "Lista: " + nombre + "\n" + "\n" + "Canciones: " + "\n" + des;
		} else {
			return "La lista: " + nombre + " está vacía\n";
		}
	}

	@Override
	public void aumentarReproducciones(){
		this.reproducciones++; 
		for (int i = 0; i < lista.size(); i++) {
			lista.get(i).aumentarReproducciones();
		}
	}
	
	public String colaborativa(ArrayList<Usuario> usuarios) { //Para que se asigne individualmente a cada usuario
		for(Usuario persona: usuarios) {
			persona.getColeccion().agregarLista(this, persona);
		}
		return "Lista Colaborativa creada";
	}
	
	public String getDescripcion() {
		return descripcion;
	}

	public void setDescripcion(String descripcion) {
		this.descripcion = descripcion;
	}
	
	public static ArrayList<Cancion> listaPorGenero(Genero genero){
		ArrayList<Cancion> cancionGenero = new ArrayList<Cancion>();
		
		for(Cancion existentes: Cancion.getCancionesDisponibles()) {
			if(existentes.getGenero().equals(genero)) {
				cancionGenero.add(existentes);
			}
		} return cancionGenero;
	}
		
	public int totalPorGenero(Genero genero) {
		
		int total = 0;
		for(Cancion cancion: getLista()) {
			if(cancion.getGenero().equals(genero)) {
				total += 1;
			}
		}
		return total;
	}
}

File: Python/src/gestorAplicacion/gestorMusica/lista.py
from gestorAplicacion.gestorMusica.musica import Musica
from gestorAplicacion.gestorMusica.cancion import Cancion
from gestorAplicacion.gestorMusica.genero import Genero

''' Módulo lista.py
    Autores: Carolina Álvarez Murillo, Miller Johan Chica Acero,
             Tomás Rodríguez Taborda, Jerónimo Ledesma Patiño,
             Catalina Restrepo Salgado
    Este módulo contiene la clase Lista, que hereda de la clase Música
'''

class Lista(Musica):

    __listasExistentes = [] #List[Lista]

    def __init__(self, str_nombre, Usuario_usuario = None, str_descripcion = "", list_lista: list = [], list_usuarios = [], set_listaColaborativa = {}):
        super().__init__(str_nombre)
        self.__usuario = Usuario_usuario
        self.__descripcion = str_descripcion
        self.__lista = list_lista
        self.usuarios = list_usuarios
        self.__listaColaborativa = set_listaColaborativa
        Lista.__listasExistentes.append(self)
    
    def duracionLista(self):
        duracion = 0
        for i in self.__lista:
            duracion += i.getDuracion()
        return duracion

    def getLista(self):
        return self.__lista #list
    
    def setLista(self, list_lista):
        self.__lista = list_lista

    def getListaColaborativa(self):
        return self.__listaColaborativa #list

    def setListaColaborativa(self, list_listaColaborativa):
        self.__listaColaborativa = list_listaColaborativa

    def getUsuario(self):
        return self.__usuario 

    def setUsuario(self, Usuario_usuario):
        self.__usuario = Usuario_usuario
    
    @classmethod
    def getListasExistentes(cls):
        return cls.__listasExistentes #list
    
    @classmethod
    def setListasExistentes(cls, list_listasExistentes):
        cls.__listasExistentes = list_listasExistentes
    
    def getUsuarios(self):
        return self.usuarios #list
    
    def setUsuarios(self, list_usuarios):
        self.usuarios = list_usuarios
    
    def agregarCancion(self, Cancion_cancion):
        if Cancion_cancion in self.__lista:
            return f"La canción {Cancion_cancion.getNombre()} ya está dentro de la lista"
        else:
            self.__lista.append(Cancion_cancion)
            return "Se ha agregado la canción " + Cancion_cancion.getNombre() + " a la lista " + self.getNombre() + " con éxito"
        
    def eliminarCancion(self, Cancion_cancion):
        if Cancion_cancion in self.__lista:
            self.__lista.remove(Cancion_cancion)
            return "Se ha eliminado la canción " + Cancion_cancion.getNombre() + " de la lista " + self.getNombre() + " con éxito"      
        else:
            return f"La canción {Cancion_cancion.getNombre()} no está dentro de la lista"

    def __str__(self):
        return "Se está reproduciendo la lista: " + self.getNombre()
    
    def infoLista(self):
        des = ""
        if len(self.__lista) > 0:
            for i in self.__lista:
                des += i.descripcion() + "\n\n"
            return "Lista: " + self.getNombre() + "\n" + "\n" + "Canciones: " + "\n" + des
        else:  
            return "La lista: " + self.getNombre() + " está vacía\n"
        
    def infoListaColaborativa(self):
        des = ""
        if len(self.__listaColaborativa) > 0:
            for i in self.__listaColaborativa:
                des += i.descripcion() + "\n\n"
            return "Lista: " + self.getNombre() + "\n" + "\n" + "Canciones: " + "\n" + des
        else:  
            return "La lista: " + self.getNombre() + " está vacía\n"

    def infoColaborativa(self):
        des = ""
        if len(self.__listaColaborativa) > 0:
            for cancion in self.__listaColaborativa:
                des += cancion.descripcion() + "\n"
            return "Lista: " + self.getNombre() + "\n" + "\n" + "Canciones: " + "\n" + des
        else:
            return "Lista: " + self.getNombre() + " está vacía"
        
    def aumentarReproducciones(self):
        self.setReproducciones(self.getReproducciones() + 1)
        for cancion in self.__lista:
            cancion.aumentarReproducciones()
    
    def colaborativa(self, list_Usuario):
        for usuario in list_Usuario:
            usuario.getColeccion().agregarLista(self, usuario)
        return "Lista colaborativa creada"
    
    def getDescripcion(self):
        return self.__descripcion #str
    
    def setDescripcion(self, str_descripcion):
        self.__descripcion = str_descripcion
    
    @classmethod
    def listaPorGenero(cls, Genero):
        lista = []
        for cancion in Cancion.getCancionesDisponibles():
            if cancion.getGenero() == Genero:
                lista.append(cancion)
        return lista

    def totalPorGenero(self, Genero_genero):
        total = 0
        for cancion in self.getLista():
            if cancion.getGenero() == Genero_genero:
                total += 1 
        return total
